smack-websocket/src/main/java/org/jivesoftware/smack/websocket/implementations/okhttp/OkHttpWebsocketImpl.java

```diff
@@ -0,0 +1,190 @@
+                              .build();
+
+        WebSocketListener listener = new WebSocketListener() {
+            @Override
+            public void onOpen(WebSocket webSocket, Response response) {
+                LOGGER.log(Level.INFO, "Websocket is open");
+                phase = WebsocketConnectionPhase.readyToSendOpen;
+            }
+            @Override
+            public void onMessage(WebSocket webSocket, String text) {
+                LOGGER.log(Level.INFO, "Received: " + text);
+                if (isCloseElement(text)) {
+                    // The server may close the connection at any time.
+                    phase = WebsocketConnectionPhase.closingHandshake;
+                    if (!closeInitiatedByClient) {
+                        // If client has not initiated the closing handshake, the handshake is initiated by the server.
+                        closeInititatedByServer = true;
+                    }
+                }
+                String closingStream = "</stream>";
+                switch (phase) {
+                case openFrameSent:
+                    if (isOpenElement(text)) {
+                        // Converts the <open> element received into <stream> element.
+                        openStreamHeader = getStreamFromOpenElement(text);
+                        phase = WebsocketConnectionPhase.readyToReceiveFeatures;
+                        try {
+                            connectionInternal.onStreamOpen(PacketParserUtils.getParserFor(openStreamHeader));
+                        } catch (XmlPullParserException | IOException e) {
+                            LOGGER.log(Level.WARNING, "Exception caught: ", e);
+                        }
+                    } else {
+                        LOGGER.log(Level.WARNING, "Unexpected Frame received", text);
+                    }
+                    break;
+                case readyToReceiveFeatures:
+                    phase = WebsocketConnectionPhase.exchangingTopLevelStreamElements;
+                    connectionInternal.parseAndProcessElement(openStreamHeader + text + closingStream);
+                    break;
+                case exchangingTopLevelStreamElements:
+                    if (isSuccessElement(text)) {
+                        phase = WebsocketConnectionPhase.readyToSendOpen;
+                    }
+                    connectionInternal.parseAndProcessElement(openStreamHeader + text + closingStream);
+                    break;
+                case closingHandshake:
+                    if (isCloseElement(text)) {
+                        connectionInternal.onStreamClosed();
+                    }
+                    break;
+                default:
+                    LOGGER.log(Level.INFO, "Default text: " + text);
+                }
+            }
+            @Override
+            public void onFailure(WebSocket webSocket, Throwable t, Response response) {
+                failureMap.put(endpoint, t);
```

Comment:
This method is written only with connection establishing in mind. But if the connection was established, then `notifyConnectionError()` should be invoked. If this was invoked due a connection attempt failure, then `setCurrentConnectionExceptionAndNotify()` should be called.

Reply:
FYI I am not sure if all those methods are already available in `ConnectionInternal`.